Added proper error handling in InputParser.jl

EmilSoleymani opened the pull request:
# Added Exception Handling in InputParser.jl
If there is anything wrong with the input file format, like missing info or data of the wrong type (i.e. a floating point number was entered instead of a whole number), corresponding errors are thrown with useful error messages as feedback. Also incorporated catching these errors and handling them as needed when calling them in `runtests.jl`.

smiths commented on test/runtests.jl:
```diff
@@ -101,7 +161,27 @@ end
+            println("Error! Skipping Tests!")
+        elseif id == Int(InputParser.SoilNumberErrorId)
+            println("Error: Invalid input file!")
+            println("\t>Line $(e.line): Soil layer number must be larger than previous one!") 
```
I see essentially the same code over and over again in this change.  Can you refactor the code so that you only have this once?

smiths reviewed (changes requested):
@EmilSoleymani, I don't see what I'm looking for in the test cases.  I might be missing it, or you might need to add a few more test cases.  I'd like test cases that tests that the exceptions are raised correctly.  In Pyunit the test case would have a syntax like @AssertRaises(ExceptionName).  I did a quick google and I found on a [testing with Julia webpage](https://docs.julialang.org/en/v1/stdlib/Test/) test syntax like this: `@test_throws exception expr`.  Do you use that (or something similar) somewhere and I missed it?

EmilSoleymani commented:
> @EmilSoleymani, I don't see what I'm looking for in the test cases. I might be missing it, or you might need to add a few more test cases. I'd like test cases that tests that the exceptions are raised correctly. In Pyunit the test case would have a syntax like @AssertRaises(ExceptionName). I did a quick google and I found on a [testing with Julia webpage](https://docs.julialang.org/en/v1/stdlib/Test/) test syntax like this: `@test_throws exception expr`. Do you use that (or something similar) somewhere and I missed it?

I never got to adding that. I made an issue about it (#64) and will be adding it shortly.

smiths commented:
Thank you for the clarification @EmilSoleymani.  I read the issue incorrectly.  You are saying what you are going to do, and for some reason I interpreted as something that you have already done.  :smile:   It doesn't hurt that we had a conversation to clarify anyway.  :smile:

EmilSoleymani commented:
I have added the test cases which confirm the correct errors are being thrown given various bad input files. This is achieved using the `@test_throws` macro as you suggested. I have also cleaned up the `runtests.jl` file. There are big chunks of code where I compare the `InputData` object and its values that were parsed from a file to its expected values. Since each file could have many different types of values, there is no elegant solution other than having a function containing all the test cases for each input file (these functions are in a module `Tests` of `tests.jl`) and having an array of these testing functions. The `runtests.jl` file now has one for loop where it goes over all test files, parses `InputData` object, and executes the corresponding test function.

smiths commented on test/runtests.jl:
```diff
@@ -7,133 +7,41 @@ using .OutputFormat
+    # Thus, I copied the exact type from above to the line below
+    println("\nTesting file 5, error expexted on line 6:")
+    @test_throws Main.OutputFormat.InputParser.ParsingError OutputData(INPUT_TEST_PATHS[5])
+    println("\nTesting file 6, error expexted.")
```
typo for expected (and in the other print statement too)

EmilSoleymani replied:
Typos fixed in latest commit

smiths commented on src/InputParser.jl:
```diff
@@ -1,18 +1,52 @@
+end
+Base.showerror(io::IO, e::ParsingError) = print(io, "Could not parse file.")
+
+# FoundationError is for when foundationOption is not read as 1 or 2
```
I know you are using the same file format as the original file, but once we get this working, we should update the file format to use a more meaningful string for the input.  Remember what 1 or 2 means is asking too much of the user.  Ideally, we'll eventually have the GUI interface which would let someone pick their foundation option with a drop down.
